### PR TITLE
fix: don't re-advertise if we have no offers

### LINKED ISSUE #160: update network offer advertisement if no offers are available
in net/discovery.go, if we have ever made and advertised offers, we continue to do that indefinitely. fix this so we don't continue to advertise if we have no offers, and update what's advertised if our offer list empties or becomes non empty.

### PR DESCRIPTION
**Pull request by noot**
closes #160

### CODE REVIEW

File: net/discovery.go [hunk 1]
```diff
@@ -120,6 +133,11 @@ func (d *discovery) advertise() {
 			d.provides = []types.ProvidesCoin{types.ProvidesXMR}
```
Comment (dimalinux) [inline net/discovery.go:133]: Github won't let me put the comment on line 119, but that line still has the TODO that this PR is closing.

File: net/discovery.go [hunk 2]
```diff
@@ -120,6 +133,11 @@ func (d *discovery) advertise() {
...
 			doAdvertise()
 		case <-time.After(ttl):
+			offers := d.offerAPI.GetOffers()
+			if len(offers) == 0 {
+				continue
+			}
+
 			doAdvertise()
```
Comment (dimalinux) [inline net/discovery.go:141]: I suspect the average reader, including myself right now, won't know the finer nuances of libp2p DHT caching.  We may need some comments somewhere in here. It makes sense to not advertise if there are no offers, assuming that we don't need to advertise to clear previous offers. Is this an optimization, or is it changing behavior?

This function would be easier to understand if the long nested function was not nested and if this function was called `advertiseLoop`.

File: net/discovery.go (hunk 1 above)
Comment (noot) [reply net/discovery.go:107]: removed

File: net/discovery.go (hunk 2 above)
Comment (noot) [reply net/discovery.go]: updated, and added comments explaining things more

**Review by dimalinux — APPROVED** [review]
Comments are great! Thanks!

File: net/discovery.go [hunk 3]
```diff
@@ -83,50 +96,65 @@ func (d *discovery) stop() error {
...
-		ttl, err = d.rd.Advertise(d.ctx, "")
-		if err != nil {
-			log.Debugf("failed to advertise in the DHT: err=%s", err)
-			ttl = tryAdvertiseTimeout
-			return
-		}
-
-		ttl = defaultAdvertiseTTL
-	}
+	ttl = d.advertise(ttl)
```
Comment (dimalinux) [inline net/discovery.go:102]: Removing line 100 and passing the constant here might be slightly more readable.

Comment (dimalinux) [reply net/discovery.go:102]: What does `ttl` stand for? It usually means "time to live" and is used for a countdown until something is discarded. This seems more like a `nextRefresh` duration.

File: net/discovery.go [hunk 4]
```diff
@@ -83,50 +96,65 @@ func (d *discovery) stop() error {
...
 			return
 		}
 	}
 }
 
+// advertise advertises that we provide XMR in the DHT.
+// note: we only advertise that we are an XMR provider, but we don't
+// advertise our specific offers.
+// to find what our offers are, peers need to send us a QueryRequest
+// over the query subprotocol.
```
Comment (dimalinux) [inline net/discovery.go:132]: Can we mention what the return value is?  It looks like we return the duration the caller should pause before advertising again.

File: net/discovery.go [hunk 5]
```diff
@@ -83,50 +96,65 @@ func (d *discovery) stop() error {
...
+func (d *discovery) advertise(ttl time.Duration) time.Duration {
+	log.Debug("advertising in the DHT...")
+	err := d.dht.Bootstrap(d.ctx)
+	if err != nil {
+		log.Warnf("failed to bootstrap DHT: err=%s", err)
+		return ttl
```
Comment (dimalinux) [inline net/discovery.go:138]: Is this the value you wanted to return? On other errors you are returning `tryAdvertiseTimeout`.

File: net/discovery.go (hunk 3 above)
Comment (dimalinux) [reply net/discovery.go:102]: Ohhh... I'm seeing where the name came from. They are abusing that name too:
https://github.com/libp2p/go-libp2p/blob/v0.21.0/p2p/discovery/routing/routing.go#L32-L37
It was probably correctly used on the DHT record, and then some programmer thought it was a good idea to use it on a value that has almost nothing in common with the time of the DHT record's expiration.

File: net/discovery.go [hunk 6]
```diff
@@ -83,50 +96,65 @@ func (d *discovery) stop() error {
...
+
+	for _, provides := range d.provides {
+		ttl, err = d.rd.Advertise(d.ctx, string(provides))
+		if err != nil {
+			log.Debugf("failed to advertise in the DHT: err=%s", err)
+			return tryAdvertiseTimeout
+		}
+	}
+
+	ttl, err = d.rd.Advertise(d.ctx, "")
```
Comment (dimalinux) [inline net/discovery.go:149]: Why do we advertise once with "XMR" on line 142 and then a second time with ""?

File: net/discovery.go (hunk 5 above)
Comment (noot) [reply net/discovery.go:138]: yeah, this is the same logic as before, but I updated it

File: net/discovery.go (hunk 6 above)
Comment (noot) [reply net/discovery.go:149]: however there isn't a loop that tries to discover new peers if our peer count is low, that should probably be implemented

Comment (dimalinux) [reply net/discovery.go]: > one says that we're in the network and allows others to discover us no matter what, the other says we're XMR providers specifically.

We only call `advertise` when we have XMR offers.  `swapcli discover` and `swapcli query-all` both take a coin. Is there anyone using this?

**Pull request merged**